### DISCUSSION TITLE
Creating multi-platform/arch Single Executable Applications (SEA)

### DISCUSSION THREAD

**Discussion by humphd** [nodejs/help#4569] [Q&A]
I'm using (and loving!) the new [node single executable applications](https://nodejs.org/api/single-executable-applications.html) feature to make a binary for my node CLI.  Such a cool and useful feature, thank you.

I now need to create the same binary for multiple platforms/archs and I'm wondering if it's possible to do this all on my Linux machine; that is, can I create a Mac or Windows binary (possibly minus the code signing stuff)?

The process _appears_ to be mostly generic:

- create the `sea-config.json` file
- turn it into a blob with `node --experimental-sea-config sea-config.json` (question: does the `node` I use here have to be the same one I'll later inject into, or can I use my system node?  Does this command do anything specific to a platform/arch or is the blob generic?)
- copy the node binary to use (I assume I could download a different node binary for another platform/arch here)
- remove the signature (not sure how I'd do this cross-platform)
- inject the blob into the binary with `npx postject`, with some tweaks per platform.  I read the code for this, and `postject` appears to be able to determine the type of binary, so I assume this part can work on any binary type?
- sign the binary (again, not sure about this for cross-platform)

Thanks for helping me understand what is and isn't possible here.  I could spin up multiple CI runners with specific platform/arch, but I'd love to be able to do this all from Linux if I can.

**Comment by liudonghua123** [comment]
I love this feature, I wrote a simple cli app named [`node-sea`](https://github.com/liudonghua123/node-sea/) to simplify the sea package work a few months ago, but it lacks of cross-compilation feature. 

I also forked [`pkg`](https://github.com/liudonghua123/pkg/) to continue support node 20/21 and later which support the simlar feature.

**Comment by humphd** [comment]
@liudonghua123 it's great that you're working on keeping this stuff going.  For the code signing bit, I recently found https://github.com/mtrojnar/osslsigncode, which might be promising.

**Comment by humphd** [comment]
https://github.com/indygreg/apple-platform-rs also looks interesting

**Comment by humphd** [comment]
So it seems like with some combination of https://gregoryszorc.com/docs/apple-codesign/main/ and https://github.com/mtrojnar/osslsigncode, it should be possible to do this all from a Linux machine.

**Comment by mkhurramali** [comment]
I have posted a webpack plugin here: https://www.npmjs.com/package/sea-webpack-plugin

This does cross-platform builds for the sea applications and has a docker you can use to do code signing.
